File: netlify/functions/process-file.js
const { Busboy } = require('@fastify/busboy');
const csv = require('csv-parser');
const PDFDocument = require('pdfkit');
const { Readable } = require('stream');

// Helper function to parse multipart form data
const parseMultipart = (event) => {
  return new Promise((resolve, reject) => {
    const busboy = new Busboy({
      headers: { 'content-type': event.headers['content-type'] || event.headers['Content-Type'] },
    });
    const result = { file: null };

    busboy.on('file', (fieldname, file) => {
      const chunks = [];
      file.on('data', (chunk) => chunks.push(chunk));
      file.on('end', () => { result.file = Buffer.concat(chunks); });
    });

    busboy.on('finish', () => resolve(result));
    busboy.on('error', (err) => reject(err));
    busboy.end(Buffer.from(event.body, 'base64'));
  });
};

// Helper function to parse CSV buffer
const parseCsv = (buffer) => {
  return new Promise((resolve, reject) => {
    const results = [];
    const stream = Readable.from(buffer);
    stream
      .pipe(csv())
      .on('data', (data) => results.push(data))
      .on('end', () => resolve(results))
      .on('error', (error) => reject(error));
  });
};

// Helper function to draw a text page
const drawTextPage = (doc, title, items) => {
  doc.addPage({ margin: 50 });
  doc.fontSize(18).font('Helvetica-Bold').text(title, { align: 'center' });
  doc.moveDown(2);
  doc.fontSize(12).font('Helvetica');
  items.forEach(item => {
    if (item.trim()) {
      doc.text(`• ${item.trim()}`, { indent: 20, paragraphGap: 10 });
    }
  });
};

exports.handler = async (event) => {
  try {
    // 1. PARSE THE UPLOADED CSV
    const { file } = await parseMultipart(event);
    if (!file) throw new Error('File not found in upload.');
    
    const data = await parseCsv(file);

    // 2. PROCESS THE DATA (similar to the Python script)
    const feedbackHeaders = Object.keys(data[0]).filter(h => h.includes('['));
    const numericData = data.map(row => {
        return feedbackHeaders.map(header => {
            const value = String(row[header]).split(':')[0];
            return parseFloat(value) || 0;
        });
    });
    
    const columnNames = ['Importance', 'Methods', 'Results', 'Discussion', 'Research Quality', 'Presentation Quality'];
    const totalScores = numericData.flat().filter(score => score > 0);
    const meanScore = totalScores.reduce((a, b) => a + b, 0) / totalScores.length;

    const keyTakeaways = data.map(row => row[Object.keys(row)[6]]).filter(Boolean);
    const strengths = data.map(row => row[Object.keys(row)[7]]).filter(Boolean);
    const improvements = data.map(row => row[Object.keys(row)[8]]).filter(Boolean);


    // 3. GENERATE THE PDF
    const pdfBuffer = await new Promise((resolve) => {
      const doc = new PDFDocument({ size: 'A4', margin: 50 });
      const buffers = [];
      doc.on('data', buffers.push.bind(buffers));
      doc.on('end', () => resolve(Buffer.concat(buffers)));

      // --- Page 1: Heatmap ---
      doc.fontSize(18).font('Helvetica-Bold').text('Seminar Feedback Heatmap', { align: 'center' });
      doc.moveDown(2);
      
      const heatmapStartY = 150;
      const cellWidth = 80;
      const cellHeight = 25;
      const maxScore = 5; // Assuming scores are out of 5 for color calculation
      
      // Draw Column Headers
      doc.fontSize(10).font('Helvetica-Bold');
      columnNames.forEach((name, i) => {
        doc.text(name.replace(' ', '\n'), 50 + i * cellWidth, heatmapStartY - 50, { width: cellWidth, align: 'center' });
      });

      // Draw Heatmap Cells
      numericData.forEach((row, rowIndex) => {
        row.forEach((score, colIndex) => {
          const x = 50 + colIndex * cellWidth;
          const y = heatmapStartY + rowIndex * cellHeight;
          const blue = (score / maxScore) * 255;
          const red = 255 - blue;
          // [R, G, B] color
          const color = [red, 0, blue]; 

          doc.rect(x, y, cellWidth, cellHeight).fill(color);
          doc.fillColor('white').text(score.toFixed(1), x, y + 8, { width: cellWidth, align: 'center' });
        });
      });
      doc.font('Helvetica').fontSize(12).fillColor('black').text('Responses', 20, heatmapStartY + (numericData.length * cellHeight) / 2, { rotation: -90 });
      
      // --- Page 2: Mean Overall Score ---
      doc.addPage();
      doc.fontSize(40).font('Helvetica-Bold').text(meanScore.toFixed(2), { align: 'center' }, doc.page.height / 2 - 50);
      doc.fontSize(20).font('Helvetica').text('Mean Overall Score', { align: 'center' });

      // --- Subsequent Pages: Qualitative Feedback ---
      drawTextPage(doc, 'Key Takeaways', keyTakeaways);
      drawTextPage(doc, 'Speaker Strengths', strengths);
      drawTextPage(doc, 'Suggestions for Improvement', improvements);
      
      doc.end();
    });

    // 4. RETURN THE PDF RESPONSE
    return {
      statusCode: 200,
      headers: {
        'Content-Type': 'application/pdf',
        'Content-Disposition': 'attachment; filename="seminar_summary.pdf"',
      },
      body: pdfBuffer.toString('base64'),
      isBase64Encoded: true,
    };

  } catch (error) {
    console.error(error);
    return {
      statusCode: 500,
      body: JSON.stringify({ error: error.message }),
    };
  }
};

File: netlify/functions/process_file.py
import base64
import cgi
import io
import json
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from matplotlib.backends.backend_pdf import PdfPages

def handler(event, context):
    # --- 1. Parse the multipart/form-data request ---
    try:
        if not event.get('isBase64Encoded', False):
            raise ValueError("Request body must be base64 encoded.")
            
        body = base64.b64decode(event['body'])
        headers = event['headers']
        
        # cgi.parse_header is crucial for getting the boundary
        ctype, pdict = cgi.parse_header(headers.get('content-type') or headers.get('Content-Type'))
        pdict['boundary'] = bytes(pdict['boundary'], "utf-8")
        
        # Parse the form data to get the file content
        form_data = cgi.parse_multipart(io.BytesIO(body), pdict)
        file_content = form_data.get('file')[0]
        
        # Read the file content into a pandas DataFrame
        csv_file = io.BytesIO(file_content)
        df = pd.read_csv(csv_file)
        
    except Exception as e:
        return {
            'statusCode': 400,
            'body': json.dumps({'error': f'Failed to parse uploaded file: {str(e)}'})
        }

    # --- 2. Run your original processing logic ---
    # Get feedback columns (those with brackets in their names)
    feedback_cols = [col for col in df.columns if '[' in col and ']' in col]
    feedback_data = df[feedback_cols]
    
    # For each column, split values at ':' and take the first part
    feedback_data = feedback_data.applymap(lambda x: str(x).split(':')[0] if isinstance(x, str) else x)
    feedback_data.columns = ['Importance', 'Methods', 'Results', 'Future_Discussion', 'Quality_of_Research', 'Quality_of_Presentation']
    feedback_data = feedback_data.apply(pd.to_numeric, errors='coerce')
    mean_score = feedback_data.mean(axis=1).mean()

    # --- 3. Generate the PDF in memory ---
    pdf_buffer = io.BytesIO()
    with PdfPages(pdf_buffer) as pdf:
        # Page 1: Heatmap
        plt.figure(figsize=(10, 8))
        sns.heatmap(feedback_data, annot=False, cmap='coolwarm', cbar=True, linewidths=0.5)
        plt.title('Seminar Feedback Heatmap', fontsize=18)
        plt.xlabel('Feedback Category', fontsize=14)
        plt.xticks(rotation=45, ha='right', fontsize=12)
        plt.ylabel('Responses', fontsize=14)
        plt.tight_layout()
        pdf.savefig()
        plt.close()

        # Page 2: Mean Overall Score
        plt.figure(figsize=(8.5, 11))
        plt.axis('off')
        plt.text(0.5, 0.5, f'Mean Overall Score: {mean_score:.2f}', ha='center', va='center', fontsize=32)
        pdf.savefig()
        plt.close()
        
        # Page 3: Key Takeaways
        plt.figure(figsize=(8.5, 11))
        plt.axis('off')
        takeaways = df.iloc[:, 6].dropna().astype(str).tolist()
        takeaways_text = "\n\n• ".join(takeaways)
        plt.text(0.05, 0.95, "Key Takeaways:\n\n• " + takeaways_text, fontsize=12, va='top', wrap=True)
        pdf.savefig()
        plt.close()

        # Page 4: Speaker Strengths
        plt.figure(figsize=(8.5, 11))
        plt.axis('off')
        strengths = df.iloc[:, 7].dropna().astype(str).tolist()
        strengths_text = "\n\n• ".join(strengths)
        plt.text(0.05, 0.95, "Speaker Strengths:\n\n• " + strengths_text, fontsize=12, va='top', wrap=True)
        pdf.savefig()
        plt.close()

        # Page 5: Suggestions for Improvement
        plt.figure(figsize=(8.5, 11))
        plt.axis('off')
        improvements = df.iloc[:, 8].dropna().astype(str).tolist()
        improvements_text = "\n\n• ".join(improvements)
        plt.text(0.05, 0.95, "Suggestions for Improvement:\n\n• " + improvements_text, fontsize=12, va='top', wrap=True)
        pdf.savefig()
        plt.close()
        
    pdf_buffer.seek(0)

    # --- 4. Return the PDF as a downloadable attachment ---
    return {
        'statusCode': 200,
        'headers': {
            'Content-Type': 'application/pdf',
            'Content-Disposition': 'attachment; filename="seminar_summary.pdf"'
        },
        'body': base64.b64encode(pdf_buffer.read()).decode('utf-8'),
        'isBase64Encoded': True
    }
